### ISSUE TITLE
Add ability to return a copy to Module's state_dict member function.

### ISSUE THREAD

**Issue by odysseus9672** [pytorch/pytorch#34765]
## 🚀 Feature

Add the ability for the `Module`'s `state_dict` member function to optionally provide detached copies of the module's internal state. Probably controlled with a keyword argument `copy=False`. 

## Motivation

When optimizing the loss for a neural network I'm working on the optimizer seems to have a tendency to overshoot or not finish converging in my iteration limit. My simple solution is to use whichever iteration has the lowest loss, but I have to keep track of that myself. I thought I could copy the state of my model using the `state_dict` function, but I've discovered that this doesn't actually copy the state, just gives me references to it. So the state keeps getting updated. I can use the `dumps` and `loads` functions from Python's `pickle` library to get around this problem, but that seems sub-optimal.

**Comment by colesbury** [comment]
A typical pattern is to save the best state_dict to a file. If you prefer keeping the values in memory, you can use `copy.deepcopy` from the standard Python library. 

I don't see a compelling enough argument for expanding the interface of `state_dict()` when there are built-in functions to perform the same operation, but I'll leave this issue open for other PyTorch contributors to comment on.
